common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactories.kt

```diff
@@ -65,3 +65,42 @@ fun JSONObject.buildCustomerInfo(): CustomerInfo {
         originalPurchaseDate = originalPurchaseDate
     )
 }
+
+/**
+ * Parses expiration dates in a JSONObject
+ * @throws [JSONException] If the json is invalid.
+ */
+private fun JSONObject.parseExpirations(): Map<String, Date?> {
+    return parseDates("expires_date")
+}
+
+/**
+ * Parses purchase dates in a JSONObject
+ * @throws [JSONException] If the json is invalid.
+ */
+private fun JSONObject.parsePurchaseDates(): Map<String, Date?> {
+    return parseDates("purchase_date")
+}
+
+/**
+ * Parses dates that match a JSON key in a JSONObject
+ * @param jsonKey Key of the dates to deserialize from the JSONObject
+ * @throws [JSONException] If the json is invalid.
+ */
+private fun JSONObject.parseDates(jsonKey: String): HashMap<String, Date?> {
+    val expirationDates = HashMap<String, Date?>()
+
+    val it = keys()
+    while (it.hasNext()) {
+        val productId = it.next()
+        val basePlanId = getJSONObject(productId).optString("product_plan_identifier")
+            .takeIf { it.isNotEmpty() }?.split(":")?.firstOrNull()
+
+        val expirationObject = getJSONObject(productId)
+
+        val key = basePlanId?.let { "$productId:$it" } ?: productId
```

Comment:
Why this concatenation here? We don't really want to force the product identifiers to be `ProductId:BasePlanId` necessarily.